### PR TITLE
Add `device` convenience function

### PR DESCRIPTION
**Pull request by mwarusz**
for getting the device on which array elements are stored.

### CODE REVIEW

**Comment by jkozdon** [comment]
bors try

**Comment by jkozdon** [comment]
Looks good to me!

File: src/GPUifyLoops.jl [hunk 1]
```diff
@@ -15,10 +15,22 @@ struct CUDA <: GPU end
 struct ROCm <: GPU end
 =#
 
-export CPU, CUDA, Device
+"""
+   device(::AbstractArray)
+
+Returns the device which stores elements of a given array.
+"""
+device(::AbstractArray) = CPU()
```
Comment (vchuravy) [inline src/GPUifyLoops.jl]:
```suggestion
device(A::AbstractArray) = error("No device associated with $(typeof(A))")
device(::Array) = CPU()
```

**Review by vchuravy — COMMENTED** [review]
Looks good to me, the only thing I would want is to be a bit more conservative.
A lot of things are `AbstractArray`s and I think it's better to give users a error,
instead of having them run into hard to understand behaviour.

**Comment by jkozdon** [comment]
So @mwarusz realized that `device` is not the best since `CUDAdrv` has a `device` function. Any thoughts on a worse name? Also, maybe we should change `Device` to something else as well? 

- `ComputeDevice`?
- `LoopDevice`?
- `Platform`?
- `ComputePlatform`?
- `SIMDPlatformDeviceThingy`?

**Comment by simonbyrne** [comment]
bump?

How about `backend`? Or calling the abstract type (e.g. `Device(a::Array)`)?

**Comment by mwarusz** [comment]
FYI GPUArrays abstract_gpu_interface.jl also exports `device(::AbstractArray)` so we probably don't want to clash with that. As for `backend`, I  see that it is used by GPUifyLoops internally ...

**Comment by vchuravy** [comment]
@mwarusz  sorry that I didn't get to this sooner. I think calling it `Backend` makes sense, especially since we want multiple backends. But that will need some more thought since we can't really force more information into the `@loop` macro.

**Pull request closed without merging**